### PR TITLE
Allow updating specific fields when updating media details

### PR DESCRIPTION
**Pull request by SiobhyB**
Related to https://github.com/wordpress-mobile/gutenberg-mobile/issues/5697

## Related PRs

* `FluxC`: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2719

## Description

At the moment, when a request is made to edit a media item, its `title`, `description`, `caption`, `alt`, and `parent_id` fields are all updated ([reference](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/5805ef65e9cebb0153c38ebf98a8f6fab646648c/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java#L615-L641)).

With this PR and the related [FluxC PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/20606), the logic is updated to allow us to specify the fields that we want to update. The reason for this change is to avoid updating more fields than necessary when saving a post and updating the unattached media items (i.e. only update the field postID).

Further background and reasoning behind making this change can be found in https://github.com/wordpress-mobile/gutenberg-mobile/issues/5697, which describes a bug in the VideoPress block that's caused by the title being set for unattached items. The iOS equivalent to this change can be found at https://github.com/wordpress-mobile/WordPress-iOS/pull/20606.

### Saving a post with unattached media items
* Create a new post.
* Add an Image block.
* Upload an image from the local device.
* Wait for the upload to finish.
* Save the post.
* Observe that the post is saved successfully.
* Go to the Media page in the browser.
* Change the View mode to Classic View _(this is needed to check if the media is attached to the post)_.
Screenshot 2023-04-26 at 13 02 45

* Click on the image added.
* Observe in the details that the media is associated with the post where it was uploaded (I.e. find the text `Uploaded to: My Post`).

###  Editing media details in the Media screen
* Go to the Media screen.
* Tap on any media item.
* Edit some of the details like the title or description.
* Observe that the media item is saved successfully.
* Check the media item in the browser and observe that changes are reflected.

### https://github.com/wordpress-mobile/gutenberg-mobile/issues/5697 is solved (VideoPress block)

> **Warning**
> There is a separate open issue related to the synchronisation of VideoPress settings [discussed here](https://github.com/wordpress-mobile/WordPress-Android/pull/18327#issuecomment-1527566516). This will be addressed in a separate PR. To avoid encountering it when testing this PR, please go through the following steps using a private WordPress.com site.

* Add VideoPress block and choose the option to upload a video from your device.
* Open the settings panel to edit the video's title and description.
* Exit the editor.
* Re-open the editor and open the settings panel.
* Verify that the updated title and description remain.

## Regression Notes
1. Potential unintended areas of impact

We are updating the class that attaches freshly uploaded media after a post is saved. As such, unintended areas of impact would be associated with media attachments/uploads.  

2. What I did to test those areas of impact (or what existing automated tests I relied on)

The same manual testing carried out for [the iOS equivalent](https://github.com/wordpress-mobile/WordPress-iOS/pull/20606) of this PR have been carried out.

In addition, the existing tests for the class being updated [here](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/test/java/org/wordpress/android/ui/uploads/PostMediaHandlerTest.kt) have also been relied upon to alert us to any regressions.

3. What automated tests I added (or what prevented me from doing so)

I believe [the existing tests](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/test/java/org/wordpress/android/ui/uploads/PostMediaHandlerTest.kt) cover the cases we need here.

<hr />

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by SiobhyB** [comment]
The initial hypothesis behind the changes in this and [the related FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2719) is outlined [here](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5697#issuecomment-1521934065). TL;DR: I attempted to fix the issue by only passing the `parent_id` field to [this function](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/5805ef65e9cebb0153c38ebf98a8f6fab646648c/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java#L621-L640) after an unattached media item is saved.  

However, this doesn't resolve the bug. As can be seen from the following screen recording, changes to the video's title and description are still discarded after saving/re-opening the post. It's necessary to wait a few seconds to see this:

https://user-images.githubusercontent.com/2998162/234956877-b85213cb-d8a2-4409-816b-8001d06a0ddb.mov

After looking through various calls in the FluxC and the WordPress Android codebases, I'm really not sure why this is happening.

@mkevins , 👋 , your name has come up a lot while researching various media-related issues. Do you perhaps have any inkling of what might be happening here? I've hit a brick wall myself, so even any suggestions or vague ideas for researching next would be most appreciated. Thank you! 🙇‍♀️

**Comment by fluiddot** [comment]
> However, this doesn't resolve the bug. As can be seen from the following screen recording, changes to the video's title and description are still discarded after saving/re-opening the post. It's necessary to wait a few seconds to see this:
> 
>  videopress.mov 
> After looking through various calls in the FluxC and the WordPress Android codebases, I'm really not sure why this is happening.

@SiobhyB I tested this PR + FluxC changes and I can't reproduce the issue you shared. In my case, updating the title and description settings are not discarded after saving the post 🤔.

https://user-images.githubusercontent.com/14905380/235131416-fc54313d-8c03-4457-ab91-c87521a8eb22.mp4

**UPDATE:** I noticed that I'm getting an error when saving the post stating that the upload couldn't be uploaded. Maybe this is the cause why I'm not getting the issue. However, seems like another important issue although not sure if it's produced by this change.

Screenshot 2023-04-28 at 13 07 46

**Comment by SiobhyB** [comment]
> @SiobhyB I tested this PR + FluxC changes and I can't reproduce the issue you shared. In my case, updating the title and description settings are not discarded after saving the post 🤔.

Oh, that is interesting. It's worth noting that I'm able to reproduce the same issue shown in my screen recording with both this PR and [the installable build](https://github.com/wordpress-mobile/WordPress-Android/pull/18174#issuecomment-1483249931) (which doesn't have these changes). 

I've been testing in both an emulator (Pixel XL, API 32) and a physical device (Pixel 6, Android 13), and my test site is a simple site with the Business plan. I'll do some more experimenting to see if I can reproduce on other devices or plans...

>  I noticed that I'm getting an error when saving the post stating that the upload couldn't be uploaded. Maybe this is the cause why I'm not getting the issue. However, seems like another important issue although not sure if it's produced by this change.

I haven't been able to replicate any issues with the upload itself. 🤔

**Comment by fluiddot** [comment]
> Oh, that is interesting. It's worth noting that I'm able to reproduce the same issue shown in my screen recording with both this PR and [the installable build](https://github.com/wordpress-mobile/WordPress-Android/pull/18174#issuecomment-1483249931) (which doesn't have these changes).
>
> I've been testing in both an emulator (Pixel XL, API 32) and a physical device (Pixel 6, Android 13), and my test site is a simple site with the Business plan. I'll do some more experimenting to see if I can reproduce on other devices or plans...

I tested again (now without the upload errors) but it still works for me 😅. I'm testing with a physical device (Samsung Galaxy S20 5G - Android 13) and a Simple site with a premium plan.

Also tried using a Pixel device via emulator but it works for me too.

https://user-images.githubusercontent.com/14905380/235156927-9a6fa4ed-29ad-42b8-adfe-b28c2d5d42d8.mp4

I noticed in the video capture you shared [here](https://github.com/wordpress-mobile/WordPress-Android/pull/18327#issuecomment-1526228675) that you're editing a published post. However, I confirm it's also working for me on published posts.

@SiobhyB it would be great if you could confirm what metadata is displaying when navigating to the video in a browser. If it contains the values you set when changing the block settings, this issue might a caching issue. On Android, the API requests sent to the host app can be cached ([reference](https://github.com/WordPress/gutenberg/blob/65adadb97f044365a0f6137cb3b2524e92a9c4c7/packages/react-native-editor/src/api-fetch-setup.js#L19C17-L23)), and by default, all requests send this parameter as `true`. You could confirm this by adding logs to the [`fetchHandler`](https://github.com/WordPress/gutenberg/blob/65adadb97f044365a0f6137cb3b2524e92a9c4c7/packages/react-native-editor/src/api-fetch-setup.js#LL28C7-L28C19) and observe the responses.

**Comment by SiobhyB** [comment]
> @SiobhyB it would be great if you could confirm what metadata is displaying when navigating to the video in a browser. If it contains the values you set when changing the block settings, this issue might a caching issue. On Android, the API requests sent to the host app can be cached ([reference](https://github.com/WordPress/gutenberg/blob/65adadb97f044365a0f6137cb3b2524e92a9c4c7/packages/react-native-editor/src/api-fetch-setup.js#L19C17-L23)), and by default, all requests send this parameter as true. You could confirm this by adding logs to the [fetchHandler](https://github.com/WordPress/gutenberg/blob/65adadb97f044365a0f6137cb3b2524e92a9c4c7/packages/react-native-editor/src/api-fetch-setup.js#LL28C7-L28C19) and observe the responses.

Ah, wow, the caching was the culprit! Thank you so much for steering me in the right direction here. I'll work on a workaround for the caching issue in a separate PR to this one. 🙇‍♀️

File: WordPress/src/main/java/org/wordpress/android/ui/uploads/PostMediaHandler.kt [hunk 1]
```diff
@@ -15,6 +15,7 @@ class PostMediaHandler
             val mediaForPost = mediaStore.getMediaForPost(post)
             mediaForPost.filter { it.postId == 0L }.forEach { media ->
                 media.postId = post.remotePostId
+                media.fieldsToUpdate = arrayOf("parent_id")
```
Comment (SiobhyB) [inline WordPress/src/main/java/org/wordpress/android/ui/uploads/PostMediaHandler.kt]: This is being added within the `updateMediaWithoutPostId` function, which is called when unattached media is uploaded to a post (background can be found in https://github.com/wordpress-mobile/WordPress-Android/pull/13269). As such, it's hoped that only updating the media item's `parent_id` is necessary here and any updates to other fields would be unnecessary.

**Comment by mkevins** [comment]
Hi @SiobhyB :wave: :smiley: 

It looks like Carlos beat me to it (though I'm not sure I would have considered caching as a culprit right away :sweat_smile: ). But, please do let me know if there's another aspect to this issue that I can help with.

**Review by jhnstn — APPROVED** [review]
The test steps with media and the VP block all worked as expected for me.
Nice work!

**Review by antonis — APPROVED** [review]
Thank you for this enhancement @SiobhyB 🙇 
Everything LGTM reviewing it along [the FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2719) 🎉

**Pull request merged**